src/address-book/AddressBook.test.tsx

```diff
@@ -0,0 +1,74 @@
+import '@testing-library/jest-dom/extend-expect'
+import React from 'react'
+import Web3 from 'web3'
+import {fireEvent, render, waitFor} from '@testing-library/react'
+import userEvent from '@testing-library/user-event'
+import {AddressBook} from './AddressBook'
+import {WalletState, WalletStatus} from '../common/wallet-state'
+import {SettingsState} from '../settings-state'
+import {BackendState} from '../common/backend-state'
+import {expectNoValidationErrorOnSubmit} from '../common/test-helpers'
+
+const web3 = new Web3()
+
+it('adds a new contact', async () => {
+  const initialState = {walletStatus: 'LOADED' as WalletStatus, web3}
+  const {queryByText, getByLabelText, getByTestId} = render(
+    <SettingsState.Provider>
+      <WalletState.Provider initialState={initialState}>
+        <BackendState.Provider initialState={{web3}}>
+          <AddressBook />
+        </BackendState.Provider>
+      </WalletState.Provider>
+    </SettingsState.Provider>,
+  )
+
+  const newContactButton = getByTestId('add-contact-button')
+  userEvent.click(newContactButton)
+
+  // dialog opened
+  await waitFor(() => expect(queryByText('Edit Contact')).toBeInTheDocument())
+
+  const saveButton = getByTestId('right-button')
+  expect(saveButton).toBeDisabled()
+
+  const cancelButton = getByTestId('left-button')
+  expect(cancelButton).toBeEnabled()
+
+  // address validation works
+  const addressInput = getByLabelText('Address')
+  const INVALID_ADDRESS_MSG = 'Invalid address'
+
+  fireEvent.change(addressInput, {target: {value: 'not an address'}})
+  await waitFor(() => expect(queryByText(INVALID_ADDRESS_MSG)).toBeInTheDocument())
+
+  const FINAL_ADDRESS = '0xffffffffffffffffffffffffffffffffffffffff'
+  fireEvent.change(addressInput, {target: {value: FINAL_ADDRESS}})
+  await waitFor(() => expect(queryByText(INVALID_ADDRESS_MSG)).not.toBeInTheDocument())
+
+  expect(saveButton).toBeDisabled()
+
+  // label validation works
+  const labelInput = getByLabelText('Label')
+  const INVALID_LABEL_MSG = 'Label length must not exceed 80 characters'
```

Comment:
Maybe we could do `Label length must not exceed ${LIMIT} characters`. Btw do we want to search by text like this in general? (If so, can we somehow use translations?) Don't we want to `getByTestId` everything?

Reply:
In general it's better to search by text as e.g. an empty element could have the right test id, but it's completely useless for the user. So we want to simulate the user interactions as closely as possible. Sometimes it isn't worth it to query by text (e.g. too cumbersome). This should be evaluated on a case-by-case basis.

The translation types already protect us from missing translations, so there's no need to "localize" the tests.
We could do the local template string though, I'm fine with changing it.